### PR TITLE
create MockPeiServices

### PR DESCRIPTION
**Pull request by VivianNK**
(cherry picked from commit 918d50d11c8ef256273090f326f9ccb8b3a21c42)

# Preface

Please ensure you have read the [contribution docs](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md) prior
to submitting the pull request. In particular,
[pull request guidelines](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md#pull-request-best-practices).

## Description

<_Please include a description of the change and why this change was made._>

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [ ] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [x] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

<_Please describe the test(s) that were run to verify the changes._>

## Integration Instructions

<_Describe how these changes should be integrated. Use N/A if nothing is required._>

### CODE REVIEW

File: MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.inf [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+[Packages]
+  MdePkg/MdePkg.dec
+  UnitTestFrameworkPkg/UnitTestFrameworkPkg.dec
+
+[LibraryClasses]
+  GoogleTestLib
+
+[BuildOptions]
+  MSFT:*_*_*_CC_FLAGS = /EHsc
```
Comment (kuqin12) [inline MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.inf:33]: Dumb question: what is this flag for?

File: MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+// This will be defined INSIDE the test, with its definition pointing to the mock function getVariable
+static EFI_PEI_READ_ONLY_VARIABLE2_PPI peiReadOnlyVariablePpi = {
+  pei_GetVariable,  // EFI_PEI_GET_VARIABLE2
+  NULL,             // EFI_PEI_GET_NEXT_VARIABLE_NAME2
+};
+
+
+extern "C" {
+  EFI_PEI_READ_ONLY_VARIABLE2_PPI* PPIVariableServices = &peiReadOnlyVariablePpi;
+}
```
Comment (kuqin12) [inline MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.cpp]: nit: new line at the end.

File: MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.cpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#include <GoogleTest/Library/MockPeiServicesLib.h>
+
+MOCK_INTERFACE_DEFINITION(MockPeiServicesLib);
+MOCK_FUNCTION_DEFINITION(MockPeiServicesLib, PeiServicesLocatePpi, 4, EFIAPI);
+MOCK_FUNCTION_DEFINITION(MockPeiServicesLib, pei_GetVariable, 6, EFIAPI);
+
+  
+// Normally PPIVariableServices is "found"
+// This will be defined INSIDE the test, with its definition pointing to the mock function getVariable
+static EFI_PEI_READ_ONLY_VARIABLE2_PPI peiReadOnlyVariablePpi = {
```
Comment (kuqin12) [inline MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.cpp]: Why do we have variable service here? I thought this is a PeiServicesLib mock?

**Review by kuqin12 — APPROVED** [review]

File: MdePkg/Test/Mock/Include/GoogleTest/Library/MockPeiServicesLib.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
+/** @file MockPeiServicesLib.h
+  Google Test mocks for PeiServicesLib
+  Copyright (c) 2022, Microsoft Corporation. All rights reserved.
```
Comment (apop5) [inline MdePkg/Test/Mock/Include/GoogleTest/Library/MockPeiServicesLib.h]:
```suggestion
Copyright (c) Microsoft Corporation.

```

File: MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.cpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
+/** @file MockPeiServicesLib.cpp
+  Google Test mocks for PeiServicesLib
+  Copyright (c) 2022, Microsoft Corporation. All rights reserved.
```
Comment (apop5) [inline MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.cpp]:
```suggestion
Copyright (c) Microsoft Corporation.
```

File: MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.inf [hunk 6]
```diff
@@ -0,0 +1,33 @@
+## @file
+# Google Test mock for PeiServicesLib
+#
+# Copyright (c) 2022, Microsoft Corporation. All rights reserved.
```
Comment (apop5) [inline MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.inf]:
```suggestion
# Copyright (c) Microsoft Corporation.
```

**Review by apop5 — APPROVED** [review]

File: MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.inf (hunk 1 above)
Comment (VivianNK) [reply MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.inf:33]: The EH stands for error handling and this flag enables C++ stack unwinding, so we can catch standard C++ exceptions. This also says the compiler assumes functions declared as extern “C” never throw a C++ exception.

https://learn.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-170

File: MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.cpp (hunk 3 above)
Comment (VivianNK) [reply MdePkg/Test/Mock/Library/GoogleTest/MockPeiServicesLib/MockPeiServicesLib.cpp]: Similarly to the mock of getVariable in the MockUefiRuntimeServisesTableLib, I'm mocking the functions PeiServices would retrieve. 
Would you suggest doing a mock of peiReadOnlyVariablePpi  in variable services?

**Pull request closed without merging**